From 874e2c8efeed9084cd065cf9ea5c0951f5afca02 Mon Sep 17 00:00:00 2001
From: Johannes Schindelin <johannes.schindelin@gmx.de>
Date: Wed, 13 Sep 2017 17:31:24 +0200
Subject: [PATCH 53/N] Fix possible segmentation fault in strnstr() on 64-bit
 systems

As of f22054c94d (Modify strnstr.c., 2017-08-30), the strnstr()
implementation was replaced by a version that segfaults (at least
sometimes) on 64-bit systems.

Without the _GNU_SOURCE constant defined, the prototype of memmem() is
missing, and as a consequence its return type defaults to int (and GCC
spits out a warning).

On 64-bit systems, the int data type is too small, though, to hold a
full char *, hence the upper 32-bit are cut off and bad things happen
due to a bogus pointer being used to access memory.

Reported as https://github.com/Alexpux/MINGW-packages/issues/2879 in
the MSYS2 project.

Cc: Sichen Zhao <1473996754@qq.com>
Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>
---
 newlib/libc/string/strnstr.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/newlib/libc/string/strnstr.c b/newlib/libc/string/strnstr.c
index ce32d0f..f6b0078 100644
--- a/newlib/libc/string/strnstr.c
+++ b/newlib/libc/string/strnstr.c
@@ -31,6 +31,7 @@ QUICKREF
 */
 
 #undef __STRICT_ANSI__
+#define _GNU_SOURCE
 #include <_ansi.h>
 #include <string.h>
 
-- 
2.9.0

